### ISSUE TITLE
Support batching again

### ISSUE THREAD

**Issue by jaydenseric** [jaydenseric/apollo-upload-client#34]
Support for batching has been temporarily dropped in the rush to support the final Apollo Client v2 API (see https://github.com/jaydenseric/apollo-upload-client/issues/33#issuecomment-342694896).

I am waiting on [`apollo-link-batch-http`](https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-batch-http) to be updated (it is still using `apollo-fetch`) so I can refer to it when writing our new batching terminating link.

**Comment by Eraldo** [comment]
@jaydenseric I don't get how to use `apollo-fetch-upload` with `apollo-link-batch-http`. It might be because I am using apollo-angular and I don't seem to get the import working.
Is there anything that I can do to clarify?

**Comment by jaydenseric** [comment]
My new advice is to wait until Apollo v2 has proper support for batching. `apollo-fetch-upload` uses v1 of the [GraphQL multipart request spec](https://github.com/jaydenseric/graphql-multipart-request-spec), which is incompatible with the latest spec for `apollo-upload-server`.

You *could* downgrade everything to use the old spec, but really you don't want to since the latest versions require totally different resolver code and work so much better that the old, now unsupported versions.

**Comment by jaydenseric** [comment]
Also @Eraldo keep an eye on https://github.com/jaydenseric/apollo-upload-client/issues/44 since you are using Angular.

**Comment by Eraldo** [comment]
Great, thank you @jaydenseric for your explanation.
Yeah.. I could not resist upgrading anyway. :D

So, I guess I will remove batching and hope it will be re-supported soon. ;)

**Comment by adamjking3** [comment]
Any news on batching?

**Comment by jaydenseric** [comment]
@adamjking3 yes! @evans [has been busy](https://github.com/apollographql/apollo-link/pull/364) adding batching support to `apollo-link-http` and deprecating `apollo-link-batch-http`.

After a lot of discussions [it has been decided](https://github.com/apollographql/apollo/blob/master/meetings/contributors-weekly/17-12-20-file-uploads.md) to add upload support to the official `apollo-link-http` package as an opt-in feature. I will prioritise a PR for that before I consider if adding batching support here is worth the effort, or if we are better off simply deprecating `apollo-upload-client` to support the official packages.

There are no plans to deprecate `apollo-upload-server`.

**Comment by glennreyes** [comment]
@jaydenseric Sounds exciting! So from your meeting notes `apollo-upload-server` will be moved to the official docs as well?

**Comment by jaydenseric** [comment]
It will be documented along with new features in `apollo-link-http`. We may or may not move the `apollo-upload-server` repo into the Apollo org; we'll see how things pan out.

**Comment by adamjking3** [comment]
@jaydenseric Do you have any idea as to when apollo-upload-client and/or batching will be released into an upcoming version of apollo-link-http?

**Comment by jaydenseric** [comment]
Once https://github.com/apollographql/apollo-link/pull/364 is merged I will create a new PR. Because that prior PR has breaking changes it is not being rushed; I'm not sure when it will be out. Hopefully soon. My PR might take a few days to put together? There needs to be Apollo documentation updates etc. so a release will depend on the wider Apollo team, which is super busy at the moment. Realistically it might take from a few days, to a few weeks from when https://github.com/apollographql/apollo-link/pull/364 is merged. If it is easy enough I will add batching support here in the meantime.

**Comment by mxstbr** [comment]
@jaydenseric that all makes sense, maybe we shouldn't be using batching after all :blush: Thank you for the update!

**Comment by PowerKiKi** [comment]
Our app performance is severely degraded when batching is not available. Especially when loading a form where there are several select boxes each trying to load their choices.

I think batching still has a lot of value when combined with a short wait time and a page doing of seemingly unrelated things. So from my point of view, I wouldn't say that interest in batching is waning, on the contrary, it is a huge gain for a negligible cost.

**Comment by Zn4rK** [comment]
I've been following this thread for a while now, and I've disabled the batching because I need file uploads via mutations. 

Since my project is not live yet, I can get by, but I have 5 initial calls to my GraphQL backend and with batching disabled I get the overhead of the transport layer for each of these calls - instead of just once.

I don't need batching when uploading files - my users can wait for the request to finish before continuing. That's not an issue for me.

The issue I'm having right now is that I can't combine `BatchHttpLink` with `apollo-upload-client`.

**Comment by PowerKiKi** [comment]
@Zn4rK this is very exactly our situation.

I understand it will take a bit more time to solve things properly and that's fine. I am willing to wait and help when I'll have time. But i just wanted to make my voice heard, that the need for batching should not be underestimated.

**Comment by Zn4rK** [comment]
This might be a bit off topic for the batching, but I've figured that my solution might help a few others in this thread.

I'm sorry for my tone in my previous reply - reading it back a few weeks later, it feels a bit rude. 
It's actually pretty easy to use `apollo-upload-client` and `BatchHttpLink` in conjunction with each other thanks to the ability to split links with [ApolloLink.split](https://github.com/apollographql/apollo-link/blob/master/docs/source/composition.md#directional-composition).

Something like this would allow the use case I described above.
```js
const isObject = node => typeof node === 'object' && node !== null;

// rough first draft, could probably be optimised in a loads of different ways.
const hasFiles = (node, found = []) => {
  Object.keys(node).forEach((key) => {
    if (!isObject(node[key]) || found.length > 0) {
      return;
    }

    if (
      (typeof File !== 'undefined' && node[key] instanceof File) ||
      (typeof Blob !== 'undefined' && node[key] instanceof Blob)
    ) {
      found.push(node[key]);
      return;
    }

    hasFiles(node[key], found);
  });

  return found.length > 0;
};

const httpLink = ApolloLink.split(
    ({ variables }) => hasFiles(variables),
    createUploadLink(options),
    new BatchHttpLink(options),
);
```

I experimented with the idea of using [extract-files](https://github.com/jaydenseric/extract-files) for the detection of files, but since it actually extracts the files from the variables object I just moved on to recursively checking the object for files instead. I'm sure my detection has some issues with FileLists etc, since I just left them out right now.

This works great for me, since I only have uploads in mutations that users triggers.

Thanks @jaydenseric for your work on this superb implementation, and the spec!

**Comment by jaydenseric** [comment]
This is probably unblocked from being worked on, now that [`apollo-link-batch-http`](https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-batch-http) has been updated.

I can't spend time on it, but someone else feel free to take a crack at it! It has the potential to be a bit complicated and/or opinionated, so I can't promise a quick turnaround for a review and merge.

**Comment by PowerKiKi** [comment]
Thanks @Zn4rK for sharing your solution. I was able to re-activate batching for most queries in no time !

**Comment by sbrichardson** [comment]
I modified @Zn4rK solution for a slightly more performant version that is working for me.

https://gist.github.com/sbrichardson/83ce25d91841bafc9365aba914007189

**Comment by PowerKiKi** [comment]
@sbrichardson thanks for sharing. The try/catch seems excessive to me. In what cases would you expect to have an error thrown when iterating JSON objects by their keys ?

**Comment by sbrichardson** [comment]
For sure, I think it can be cleaned up some, just overly cautious, since I hadn't tested it much when I shared, and while loops crash a browser so quickly.

Side note --

I think taking advantage of context to control this process when needed is much more performant without needing to iterate over all the variables on every request. This method is especially useful since I only have a few Upload mutations, compared to many normal mutations/queries.

I've confirmed this works well. Example:

```jsx
/**
 * Apollo Terminating link
 * The split will skip Batching if an
 * operation's context contains hasUpload: true.
 */
const httpLink = ApolloLink.split(
  operation => operation.getContext().hasUpload,
  createUploadLink(OPTS),
  new BatchHttpLink(OPTS)
)

/**
 * Then within a mutation component, specify the context
 * with hasUpload to activate the Upload link.
 */
const { data } = await uploadMutation({
  variables: { file },
  context: {
    hasUpload: true, // activate Upload link
  }
})

```

**Comment by karensg** [comment]
Heej @sbrichardson,

A much cleaner solution indeed. I borrowed your idea for the context variable and published an article how to do batching/uploading with Apollo & Rails:
https://medium.com/@karensgrigorjancs/effortless-file-uploads-with-activestorage-apollo-and-react-7a50929c40ca

Might be interesting for people in this issue.
K

**Comment by digeomel** [comment]
I'm not sure if this is relevant or off-topic, but I managed to get batched upload working with Apollo Angular and the `apollo-upload-client`, without the `apollo-link-batch-http` module, because our GraphQL server (Java-based) doesn't support batching anyway.
So what I did is I just iterate over the multiple files and build a mutation with multiple calls to upload each file, something like this:
```
mutation (file1: Upload!, file2: Upload!) {
    upload1: uploadOneFile(file: file1) {
        id
        fileName
    }
    upload2: uploadOneFile(file: file2) {
        id
        fileName
    }
}
```
This is still one mutation, and one http call. Maybe this will help someone.

**Comment by MakhBeth** [comment]
thankyou @digeomel 
Your solution worked like a charm

I only had to update the mutation hooks like 
```const [upload, { data, loading, error }] = useMutation(
    mutation,
    {
      context: { hasUpload: true },
    }
  )```

**Comment by MakhBeth** [comment]
thankyou all for the split solution

Worked like a charm

I only had to update the mutation hooks 
```js
const [upload, { data, loading, error }] = useMutation(
    mutation,
    {
      context: { hasUpload: true },
    }
  )
```

**Comment by McPo** [comment]
Cheers @sbrichardson for the fix

Theres a potential modification which gets rid of setting the context (Although it might still be better to explicitly set the context)
```
/*
  Upload and batch link, are both terminating links
  Which do not support each other.
  We therefore only use the upload link when one of the variables are of type Upload
  Otherwise we use the batch link
  Alternatively we could have accomplished this by placing a boolean variable in the upload context instead
*/
const uploadAndBatchHTTPLink = opts => ApolloLink.split(
  operation => Boolean(operation.query.definitions.find(d => d.variableDefinitions.find(vd => vd.type.type?.name.value === 'Upload'))),
  createUploadLink(opts),
  new BatchHttpLink(opts)
);
```

It looks for variable definitions of type `Upload`, and if it finds any, it uses the upload link instead the batch link.

Appears to be working for me so far.

**Comment by jaydenseric** [comment]
Note that [`extract-files`](https://npm.im/extract-files) (already an `apollo-upload-client` dependency) since the past few versions doesn't mutate the input, so you could detect files in variables like this:

```js
const { extractFiles } = require('extract-files');
```

```js
const { files } = extractFiles(variables);
const hasFiles = !!files.size
```

**Comment by jaydenseric** [comment]
I don’t plan to implement support for batching at this time, please refer to the earlier discussion in this issue:

- Implementing it would be a fair bit of work. Over 2 years has passed since [I invited the community to submit a PR](https://github.com/jaydenseric/apollo-upload-client/issues/34#issuecomment-376818641), and so far no one has stepped up to the plate.
- It would increase the maintenance burden of this project.
- Please don’t mistake closing this issue for laziness; when I believe something has technical merit I move mountains to make it happen. A long time ago [I changed my mind that batching was a good idea](https://github.com/jaydenseric/apollo-upload-client/issues/34#issuecomment-358830808), and nothing dramatic has caused me to change my mind back again in the years since. I’ve created and worked on numerous apps and APIs at commercial scale without it, no problems.
- If you’re a fan of batching, you can get it to conditionally work for non file upload operations with [a little manual effort to compose links](https://github.com/jaydenseric/apollo-upload-client/issues/34#issuecomment-638901324).
